test/test_type_promotion.py

```diff
@@ -430,6 +431,98 @@ def test_transpose(self, device):
+        err = "{} {}({}, {})".format("  coalesced" if coalesced else "uncoalesced", op_name + suffix, dtype, dtype)
+
+        def op(t1, t2):
+            return getattr(t1, op_name + suffix)(t2)
+
+        add_sub = op_name == 'add' or op_name == 'sub'
+
+        (dense1, sparse1) = self._test_sparse_op_input_tensors(device, dtype, coalesced)
+        (dense2, sparse2) = self._test_sparse_op_input_tensors(device, dtype, coalesced, op_name != 'div')
+        common_dtype = torch.result_type(dense1, dense2)
+        if self.device_type == 'cpu' and common_dtype == torch.half:
+            self.assertRaises(RuntimeError, lambda: op(s1, d2))
+
+        expected = op(dense1.clone(), dense2)
+        precision = self._get_precision(expected.dtype, coalesced)
+        test_tensors = [expected, dense1, sparse1, dense2, sparse2]
+        e, d1, s1, d2, s2 = [x.clone() for x in test_tensors] if inplace else test_tensors
+
+        # Test op(sparse, sparse)
+        if op_name != 'div':
+            sparse = op(s1, s2)
+            self.assertEqual(sparse.dtype, e.dtype)
+            self.assertEqual(e, sparse.to_dense(), prec=precision, message=err)
+        else:
+            # sparse division only supports division by a scalar
+            self.assertRaises(RuntimeError, lambda: op(s1, s2).to_dense())
+
+        # Test op(dense, sparse)
+        if add_sub:
+            if inplace:
+                e, d1, s1, d2, s2 = [x.clone() for x in test_tensors]
+            dense_sparse = op(d1, s2)
+            self.assertEqual(e, dense_sparse, prec=precision, message=err)
+        else:
+            # sparse division only supports division by a scalar
+            # mul: Didn't find kernel to dispatch to for operator 'aten::_nnz'
+            self.assertRaises(RuntimeError, lambda: op(d1, s2))
+
+        # Test op(sparse, dense) not supported for any ops:
+        # add(sparse, dense) is not supported. Use add(dense, sparse) instead.
+        # sparse division only supports division by a scalar
+        # mul: Didn't find kernel to dispatch to for operator 'aten::_nnz'.
+        self.assertRaises(RuntimeError, lambda: op(s1, d2))
+
+        # Test op(sparse, scalar)
+        if not add_sub and not (self.device_type == 'cpu' and dtype == torch.half):
+            if inplace:
+                e, d1, s1, d2, s2 = [x.clone() for x in test_tensors]
+            scalar = d2.view(d2.numel())[0].item()
+
+            sparse = op(s1, scalar)
+            dense_scalar = op(d1, scalar)
+            self.assertEqual(sparse.dtype, dense_scalar.dtype)
+            self.assertEqual(dense_scalar, sparse.to_dense(), prec=precision, message=err)
+        else:
+            # add(sparse, dense) is not supported. Use add(dense, sparse) instead.
+            # "mul_cpu" / "div_cpu" not implemented for 'Half'
+            self.assertRaises(RuntimeError, lambda: op(s1, d2.view(d2.numel())[0].item()))
+
+    @onlyOnCPUAndCUDA
+    def test_sparse_ops(self, device):
```

Comment:
Can this just be `test_sparse_ops(self, device, dtype)`? I think Ruberry's test design allows for it: https://github.com/pytorch/pytorch/blob/3990e9d1ca23e39bfa65fe1d907ddb9f9dbf0919/test/test_torch.py#L6996

Reply:
sort-of, it would support it with a `@dtypes(...a very long list...)`. I don't think we have a way to generically add all math types (and the list of all math types could change). 

Also in the next diff I'm adding a second dtype for mixed-types -- I think it'd be confusing to have one dtype come through annotation but the second come via a for loop.

Reply:
Oh right, I keep on forgetting the other diff is stacked on top of this one.